File: 2023/UICPC Regional/Round #12/solutions/raffle/per.cc
#include <bits/stdc++.h>

int main(void) {
  int n, p;
  scanf("%d%d", &n, &p);
  long double best = (long double)1.0*p/(n+1);
  long double prod = (long double)p/(n+1);
  int bestm = 1;
  for (int m = 2; m <= 2*n; ++m) {
    prod *= (long double)1.0*(n-p+m)/(n+m);
    long double val = prod*m;
    if (val > best) {
      best = val;
      bestm = m;
    } 
  }
  printf("%.11Lf\n", best);
}

File: 2023/UICPC Regional/Round #12/solutions/raffle/per.php
<?php

fscanf(STDIN, "%d %d", $n, $p);
$best = 1.0*$p/($n+1);
$prod = $best;
for ($m = 2; $m <= $n+1; ++$m) {
    $prod *= 1.0*($n-$p+$m)/($n+$m);
    $val = $prod*$m;
    if ($val > $best)
       $best = $val;
    else
       break;
}
echo $best;
